### ISSUE TITLE
Feature discussion -- Providing steps information from Marathon into Allure's reports

### ISSUE THREAD

**Issue by PStrelchenko** [MarathonLabs/marathon#290]
## Providing steps information from Marathon into Allure's reports

* **Type**: Design proposal
* **Status**: Discussion

## Abstract

This is a description of new solution for adding steps information into Allure's reports generated by Marahton. 

Goals:

- Extend helpfullnes of Allure's reports by adding steps information. It looks [like this](https://demo.qameta.io/allure/#suites/fd187460706bbe14b8a84cc98c3fb991/78c6390fe4742f4f/)
- Create single contract for implementing steps results support both for Android and iOS
- Define new transport layer for delivering steps information for Android instrumentation tests

Adding steps results is very usefull for analysing problems in tests.

## Table of Contents

* [Existing solutions](#existing-solutions)
* [Proposed solution](#proposed-solution)
  * [TL;DR;](#tl-dr)
  * [Overview](#overview)
  * [Contract for steps](#contract-for-steps)
  * [Contract for steps producing and consuming](#contract-for-steps-producing-and-consuming)
  * [Transport layer discussion](#transport-layer-discussion)
    * [Overview of INSTRUMENTATION_STATUS](#overview-of-instrumentation_status)
  * [Implementation details](#implementation-details)
    * [Providing steps from third-party frameworks](#providing-steps-from-third-party-frameworks)
    * [Catching information on the Marathon's side](#catching-information-on-the-marathons-side)
* [Already asked questions (AAQ)](#already-asked-questions)

## Existing solutions

Right now we have working solution in [PR # 284](https://github.com/Malinskiy/marathon/pull/284), but it has several problems in it's implementation:

* It is bounded to a specific third-party library - [Kaspresso](https://github.com/KasperskyLab/Kaspresso), which impairs the flexibility of this solution
* For Android tests it uses LogCat as transport layer for delivering steps information (we [can't really rely on this](https://github.com/Malinskiy/marathon/issues/251))
* If test failed, there are no information about steps
* Has no specified contract to be implemented by iOS vendors

### TL;DR;

- Declare steps contract models
- Declare steps producing/consuming interfaces
- On third-party frameworks side send information into `StepsResultsProducer` which will write steps JSON into adb shell's output
- On Marathon's side fetch information from ddmlib's `testMetrics` and propagate steps information into `AllureReporter`

### Overview

Long story short - we want to propagate steps information from third-party frameworks through key-value pair with `INSTRUMENTATION_STATUS` for every test. This pair [will be parsed by ddmlib into testMetrics](https://android.googlesource.com/platform/tools/base/+/master/ddmlib/src/main/java/com/android/ddmlib/testrunner/InstrumentationResultParser.java#358) assosiated with every `TestResult`. Marathon will catch this testMetrics and then propagate them into Allure's reports.

### Contract for steps

First of all let's define some model for holding steps:

```kotlin
data class StepResult(
  val description: String,
  val startTime: Long,
  val stopTime: Long,
  val trace: String?,
  val status: StepResultStatus,
  val subSteps: List<StepResult>
)

enum class StepResultStatus {
  BROKEN,
  FAILURE,
  PASSED
}
```

`StepResult` - recursive structure, where every step can hold substeps set. Also every `StepResult` holds description, start and stop time, status - is current step finished successfully or not, and optionally - error stacktrace, if step has `failed` or `broken` status.

Steps will be attached to every individual test result (Marathon's `TestResult`): 

```kotlin
data class TestResult(
  val test: Test,
  val device: DeviceInfo,
  val status: TestStatus,
  val startTime: Long,
  val endTime: Long,
  val stacktrace: String? = null,
  val attachments: List<Attachment> = emptyList(),
  val stepsResults: List<StepResult> = emptyList()
)
```

Marathon by itself will propagate `TestResult`s into `ExecutionReport` and then into `AllureReporter` by itself.

### Contract for steps producing and consuming

For providing steps information from third-party libraries into some transport layer: 

```kotlin
interface StepsResultsProducer {
  fun produce(testClassName: String, testMethodName: String, stepsResults: List<StepResult>)
}
```

For propagating steps information from transport layer into Marathon's internals we introduce two interfaces (similar to [`AttachmentProvider` & `AttachmentListener`](https://github.com/Malinskiy/marathon/blob/develop/core/src/main/kotlin/com/malinskiy/marathon/report/attachment/AttachmentProvider.kt)):

```kotlin
interface StepsResultsProvider {
    fun registerListener(listener: StepsResultsListener)
}

interface StepsResultsListener {
    fun onStepsResults(test: Test, stepsResults: List<StepResult>)
}
```

### Transport layer discussion

As we already discussed in [PR # 284](https://github.com/Malinskiy/marathon/pull/284), LogCat used in existing solution is not reliable transport layer. But @Malinskiy noted that 

> [unknown test events are treated as metrics](https://android.googlesource.com/platform/tools/base/+/master/ddmlib/src/main/java/com/android/ddmlib/testrunner/InstrumentationResultParser.java#358)

and it can be more reliable solution. 

#### Overview of INSTRUMENTATION_STATUS

While instrumentation tests are executing, they sends some adb shell commands, which writes special strings into shell's output. For example, when some test is started, we can see in shell's output the following strings:

```
11:12:52 V/InstrumentationResultParser: INSTRUMENTATION_STATUS: class=com.kaspersky.kaspressample.simple_tests.SimpleTest
11:12:52 V/InstrumentationResultParser: INSTRUMENTATION_STATUS: current=1
11:12:52 V/InstrumentationResultParser: INSTRUMENTATION_STATUS: id=AndroidJUnitRunner
11:12:52 V/InstrumentationResultParser: INSTRUMENTATION_STATUS: numtests=1
11:12:52 V/InstrumentationResultParser: INSTRUMENTATION_STATUS: stream=
11:12:52 V/InstrumentationResultParser: com.kaspersky.kaspressample.simple_tests.SimpleTest
11:12:52 V/InstrumentationResultParser: INSTRUMENTATION_STATUS: test=test
Starting 1 tests on Pixel_2_API_28(AVD) - 9
11:12:52 V/InstrumentationResultParser: INSTRUMENTATION_STATUS_CODE: 1
```

ddmlib can parse such output to produce and propagating tests results. As we mentioned before, "unknown test events are treated as metrics", so our goal is - somehow write test steps information into shell's output. 

Every instrumentation test in Android can send status information into shell's output with invoking

```kotlin
InstrumentationRegistry.getInstrumentation().sendStatus()
```

method. But it requires specifing test identifier to be accepted by `ddmlib`. Test identifier is a pair of test class name and test method name. For example:

```kotlin
package com.kaspersky.kaspressample.simple_tests

@RunWith(AndroidJUnit::class)
class SimpleTest : TestCase() {
  
  @Test
  fun test() {
    // ...
  }


}
```

Here, test identifier == (com.kaspersky.kaspressample.simple_tests.SimpleTest + test). We can create a special model for it:

```kotlin
data class TestIdentifier(
  val className: String,
  val methodName: String
)
```

#### Providing steps from third-party frameworks

Third-party frameworks can send their steps information for Marathon with [jUnit's TestWatcher](https://junit.org/junit4/javadoc/4.12/org/junit/rules/TestWatcher.html). It can be used as [jUnit's Rule](https://github.com/junit-team/junit4/wiki/rules) and has ability to catch such event as test starting, failing, ignoring and success completing. 
Following `TestWatcher` can be used as default implementation of `StepsResultsProducer` interface to support sending information from third-party frameworks into adb shell's output.

```kotlin
class StepsResultsProducerImpl : TestWatcher(), StepsResultsProducer {

    companion object {
        private const val INSTRUMENTATION_KEY_STEPS_RESULTS_JSON = "marathon.stepsResultsJson"
        private const val EMPTY_STEPS_RESULTS_JSON = "[]"
    }


    private val stepsResultsJsonMap = mutableMapOf<TestIdentifier, String>()


    override fun produce(testClassName: String, testMethodName: String, stepsResultsJson: String) {
        val testIdentifier = TestIdentifier(testClassName, testMethodName)
        stepsResultsJsonMap[testIdentifier] = stepsResultsJson
    }


    override fun failed(e: Throwable, description: Description) {
        super.failed(e, description)
        sendStatus(REPORT_VALUE_RESULT_FAILURE, description)
    }

    override fun skipped(e: AssumptionViolatedException, description: Description) {
        super.skipped(e, description)
        sendStatus(REPORT_VALUE_RESULT_ASSUMPTION_FAILURE, description)
    }

    override fun succeeded(description: Description) {
        super.succeeded(description)
        sendStatus(REPORT_VALUE_RESULT_OK, description)
    }


    private fun sendStatus(code: Int, description: Description) {
        val stepsResultsJson = stepsResultsJsonMap[description.toTestIdentifier()]
            ?: EMPTY_STEPS_RESULTS_JSON
        InstrumentationRegistry.getInstrumentation()
            .sendStatus(code, createBundle(description, stepsResultsJson))
    }

    private fun createBundle(description: Description, stepsResultsJson: String): Bundle {
        return Bundle().apply {
            putString(REPORT_KEY_NAME_CLASS, description.className)
            putString(REPORT_KEY_NAME_TEST, description.methodName)
            putString(INSTRUMENTATION_KEY_STEPS_RESULTS_JSON, stepsResultsJson)
        }
    }


}
```

#### Catching information on the Marathon's side

Catching information from `ddmlib` can be performed with existing interface - `AndroidTestRunListener` and it's method - `testEnded(test: TestIdentifier, testMetrics: Map<String, String>)`.

```kotlin
class StepsResultsProviderRunListener(
    private val stepResultConverter: StepResultConverter
) : NoOpTestRunListener(), StepsResultsProvider {

    companion object {
        private const val INSTRUMENTATION_STATUS_KEY_STEPS_RESULTS_JSON = "marathon.stepsResultsJson"
        private const val EMPTY_STEPS_RESULTS_JSON = "[]"
    }

    private val stepsResultsListeners = mutableListOf<StepsResultsListener>()

    override fun registerListener(listener: StepsResultsListener) {
        stepsResultsListeners.add(listener)
    }


    override fun testEnded(test: TestIdentifier, testMetrics: Map<String, String>) {
        val stepsResultsJson = testMetrics[INSTRUMENTATION_STATUS_KEY_STEPS_RESULTS_JSON] ?: EMPTY_STEPS_RESULTS_JSON
        val stepsResults = stepResultConverter.fromJson(stepsResultsJson)

        stepsResultsListeners.forEach { it.onStepsResults(test.toTest(), stepsResults) }
    }

}
```

This `AndroidTestRunListener` can propagate information fetched from ddmlib's `testMetrics` into `TestRunResultsListener` which will accumulate all test results from current tests batch. 

```kotlin
class TestRunResultsListener(
    ...
    stepsResultsProviders: List<StepsResultsProvider>
) : AbstractTestRunResultListener(), AttachmentListener, StepsResultsListener {

    ...
    private val stepsResultsMap: MutableMap<Test, MutableList<StepResult>> = mutableMapOf()

    init {
        ...
        stepsResultsProviders.forEach {
            it.registerListener(this)
        }
    }

    ...

    override fun onStepsResults(test: Test, stepsResults: List<StepResult>) {
        val list = stepsResultsMap[test]
        if (list == null) {
            stepsResultsMap[test] = mutableListOf()
        }

        stepsResultsMap[test]!!.addAll(stepsResults)
    }

    ...

}
```

After propagating step results into `TestRunResultsListener` we will have avility to propagate them into `TestResult` model:

```kotlin
class TestRunResultsListener {
  
  ...

  private fun Map.Entry<TestIdentifier, AndroidTestResult>.toTestResult(device: Device): TestResult {
      ...
      val stepsResults = stepsResultsMap[test] ?: emptyList<StepResult>()
      return TestResult(
          ...
          stepResults = stepsResults
      )
  }

}
```

That's all, after this manipulations, we will fetch step results in `AllureReporter` and convert them into Allure's models. 

## Already asked questions

> Q: Do we really need recursive structure of steps?

A: Yeap, we required such structure because of some testing scenarios, where we can use nested scenarios with multiple steps.
